disabling retryOnConnectionFailure causes java.io.EOFException: \n not found: limit=0 content=…

https://github.com/spring-projects/spring-ai/blame/c52f675445cb563a9f37c54f3489e3a6e049db97/models/spring-ai-openai/src/main/java/org/springframework/ai/openai/http/okhttp/SpringAiOpenAiHttpClient.java#L516

Every time when I make several requests to llama.cpp server, I catch exception on second/third request

```
java.util.concurrent.CompletionException: com.openai.errors.OpenAIIoException: Request failed
	at java.base/java.util.concurrent.CompletableFuture.wrapInCompletionException(CompletableFuture.java:323) ~[na:na]
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:359) ~[na:na]
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:364) ~[na:na]
	at java.base/java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:957) ~[na:na]
	at java.base/java.util.concurrent.CompletableFuture$UniHandle.tryFire(CompletableFuture.java:932) ~[na:na]
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:531) ~[na:na]
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2221) ~[na:na]
	at org.springframework.ai.openai.http.okhttp.SpringAiOpenAiHttpClient$1.onFailure(SpringAiOpenAiHttpClient.java:138) ~[spring-ai-openai-2.0.0-RC1.jar!/:2.0.0-RC1]
	at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:525) ~[okhttp-4.12.0.jar!/:na]
	at io.micrometer.context.ContextSnapshot.lambda$wrap$0(ContextSnapshot.java:91) ~[context-propagation-1.2.1.jar!/:1.2.1]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1090) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:614) ~[na:na]
	at java.base/java.lang.Thread.run(Thread.java:1474) ~[na:na]
Caused by: com.openai.errors.OpenAIIoException: Request failed
	... 6 common frames omitted
Caused by: java.io.IOException: unexpected end of stream on http://llama_server:8080/...
	at okhttp3.internal.http1.Http1ExchangeCodec.readResponseHeaders(Http1ExchangeCodec.kt:210) ~[okhttp-4.12.0.jar!/:na]
	at okhttp3.internal.connection.Exchange.readResponseHeaders(Exchange.kt:110) ~[okhttp-4.12.0.jar!/:na]
	at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.kt:93) ~[okhttp-4.12.0.jar!/:na]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109) ~[okhttp-4.12.0.jar!/:na]
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.kt:34) ~[okhttp-4.12.0.jar!/:na]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109) ~[okhttp-4.12.0.jar!/:na]
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.kt:95) ~[okhttp-4.12.0.jar!/:na]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109) ~[okhttp-4.12.0.jar!/:na]
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.kt:83) ~[okhttp-4.12.0.jar!/:na]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109) ~[okhttp-4.12.0.jar!/:na]
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.kt:76) ~[okhttp-4.12.0.jar!/:na]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109) ~[okhttp-4.12.0.jar!/:na]
	at io.micrometer.core.instrument.binder.okhttp3.OkHttpObservationInterceptor.intercept(OkHttpObservationInterceptor.java:97) ~[micrometer-core-1.16.4.jar!/:1.16.4]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109) ~[okhttp-4.12.0.jar!/:na]
	at okhttp3.internal.connection.RealCall.getResponseWithInterceptorChain$okhttp(RealCall.kt:201) ~[okhttp-4.12.0.jar!/:na]
	at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:517) ~[okhttp-4.12.0.jar!/:na]
	... 4 common frames omitted
Caused by: java.io.EOFException: \n not found: limit=0 content=…
	at okio.RealBufferedSource.readUtf8LineStrict(RealBufferedSource.kt:335) ~[okio-jvm-3.6.0.jar!/:na]
	at okhttp3.internal.http1.HeadersReader.readLine(HeadersReader.kt:29) ~[okhttp-4.12.0.jar!/:na]
	at okhttp3.internal.http1.Http1ExchangeCodec.readResponseHeaders(Http1ExchangeCodec.kt:180) ~[okhttp-4.12.0.jar!/:na]
	... 19 common frames omitted
```

--- a/models/spring-ai-openai/src/main/java/org/springframework/ai/openai/http/okhttp/SpringAiOpenAiHttpClient.java
+++ b/models/spring-ai-openai/src/main/java/org/springframework/ai/openai/http/okhttp/SpringAiOpenAiHttpClient.java
@@ -513,8 +513,9 @@ public final class SpringAiOpenAiHttpClient implements HttpClient {
 
 		public SpringAiOpenAiHttpClient build() {
 			OkHttpClient.Builder okBuilder = new OkHttpClient.Builder()
-				// SDK's RetryingHttpClient owns retries; disable here to avoid doubling.
-				.retryOnConnectionFailure(false)
+				// Recover from stale pooled connections (OkHttp's default); distinct from
+				// the SDK's status-code/backoff retries, so no duplication. See gh-6318.
+				.retryOnConnectionFailure(true)
 				.pingInterval(Duration.ofMinutes(1))
 				.connectTimeout(this.timeout.connect())
 				.readTimeout(this.timeout.read())

--- a/models/spring-ai-openai/src/test/java/org/springframework/ai/openai/http/okhttp/SpringAiOpenAiHttpClientTests.java
+++ b/models/spring-ai-openai/src/test/java/org/springframework/ai/openai/http/okhttp/SpringAiOpenAiHttpClientTests.java
@@ -1,0 +1,89 @@
+/*
+ * Copyright 2023-present the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.ai.openai.http.okhttp;
+
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.RequestBody;
+import okhttp3.Response;
+import okhttp3.mockwebserver.MockResponse;
+import okhttp3.mockwebserver.MockWebServer;
+import okhttp3.mockwebserver.SocketPolicy;
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * Tests for {@link SpringAiOpenAiHttpClient} connection handling.
+ *
+ * @author Siarhei Dudzin
+ */
+class SpringAiOpenAiHttpClientTests {
+
+	private static final String CHAT_COMPLETION_RESPONSE = """
+			{"id":"chatcmpl-test","object":"chat.completion","created":1780704000,"model":"gpt-5.4",\
+			"choices":[{"index":0,"message":{"role":"assistant","content":"Hello"},"finish_reason":"stop"}],\
+			"usage":{"prompt_tokens":5,"completion_tokens":1,"total_tokens":6}}""";
+
+	@Test
+	void retryOnConnectionFailureIsEnabledByDefault() {
+		// OkHttp's transparent recovery from broken/stale pooled connections must stay on
+		// so that servers which close idle keep-alive connections (e.g. llama.cpp) do not
+		// surface "unexpected end of stream" / EOFException to callers. See gh-6318.
+		OkHttpClient client = SpringAiOpenAiHttpClient.builder().build().getOkHttpClient();
+
+		assertThat(client.retryOnConnectionFailure()).isTrue();
+	}
+
+	@Test
+	void recoversWhenReusedConnectionWasClosedByServer() throws Exception {
+		try (MockWebServer server = new MockWebServer()) {
+			// First response closes the keep-alive connection, leaving a stale entry in
+			// OkHttp's pool; the second request must transparently recover (gh-6318).
+			server.enqueue(new MockResponse().setResponseCode(200)
+				.setHeader("Content-Type", "application/json")
+				.setBody(CHAT_COMPLETION_RESPONSE)
+				.setSocketPolicy(SocketPolicy.DISCONNECT_AT_END));
+			server.enqueue(new MockResponse().setResponseCode(200)
+				.setHeader("Content-Type", "application/json")
+				.setBody(CHAT_COMPLETION_RESPONSE));
+			server.start();
+
+			OkHttpClient client = SpringAiOpenAiHttpClient.builder().build().getOkHttpClient();
+
+			// Pin to the bound loopback address: "localhost" may resolve to both
+			// 127.0.0.1 and ::1, and a retry against the address family MockWebServer is
+			// not bound to would fail with a spurious "connection refused".
+			String url = "http://127.0.0.1:" + server.getPort() + "/v1/chat/completions";
+
+			try (Response first = client.newCall(newPost(url)).execute()) {
+				assertThat(first.code()).isEqualTo(200);
+				// Drain the body so the connection is eligible for pooling/reuse.
+				first.body().string();
+			}
+
+			try (Response second = client.newCall(newPost(url)).execute()) {
+				assertThat(second.code()).isEqualTo(200);
+			}
+		}
+	}
+
+	private static Request newPost(String url) {
+		return new Request.Builder().url(url).post(RequestBody.create("{}", null)).build();
+	}
+
+}

